sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushFoldableIntoBranchesSuite.scala

```diff
@@ -259,6 +261,79 @@ class PushFoldableIntoBranchesSuite
       CaseWhen(Seq((a, FalseLiteral), (c, FalseLiteral)), None))
   }
 
+  test("SPARK-33848: Push down dateTimeExpression with binary expression through If/CaseWhen") {
+    val d = Date.valueOf("2021-01-01")
+    // If
+    assertEquivalent(AddMonths(Literal(d),
+      If(a, Literal(1), Literal(2))),
+      If(a, Literal(Date.valueOf("2021-02-01")), Literal(Date.valueOf("2021-03-01"))))
+    assertEquivalent(DateAdd(Literal(d),
+      If(a, Literal(1), Literal(2))),
+      If(a, Literal(Date.valueOf("2021-01-02")), Literal(Date.valueOf("2021-01-03"))))
+    assertEquivalent(DateAddInterval(Literal(d),
+      If(a, Literal(new CalendarInterval(1, 1, 0)),
+        Literal(new CalendarInterval(1, 2, 0)))),
+      If(a, Literal(Date.valueOf("2021-02-02")), Literal(Date.valueOf("2021-02-03"))))
+    assertEquivalent(DateAddYMInterval(Literal(d),
+      If(a, Literal.create(Period.ofMonths(1), YearMonthIntervalType),
+        Literal.create(Period.ofMonths(2), YearMonthIntervalType))),
+      If(a, Literal(Date.valueOf("2021-02-01")), Literal(Date.valueOf("2021-03-01"))))
+    assertEquivalent(DateDiff(Literal(d),
+      If(a, Literal(Date.valueOf("2021-02-01")), Literal(Date.valueOf("2021-03-01")))),
+      If(a, Literal(-31), Literal(-59)))
+    assertEquivalent(DateSub(Literal(d),
+      If(a, Literal(1), Literal(2))),
+      If(a, Literal(Date.valueOf("2020-12-31")), Literal(Date.valueOf("2020-12-30"))))
+    assertEquivalent(TimestampAddYMInterval(Literal(d.getTime, TimestampType),
+      If(a, Literal.create(Period.ofMonths(1), YearMonthIntervalType),
+        Literal.create(Period.ofMonths(2), YearMonthIntervalType))),
+      If(a, Literal.create(Timestamp.valueOf("1970-02-19 07:04:48"), TimestampType),
```

Comment:
hmm, the result should be `2021-01-01` + 1 month = `2021-02-01`.

Reply:
> hmm, the result should be `2021-01-01` + 1 month = `2021-02-01`.

Fix it